### PR TITLE
Make GH-324 work with Java 7 and 6

### PR DESCRIPTION
**Pull request by noprompt**
This patch makes a few tweaks to the work done in GH-324 by refactoring the parameter name derivation and accounting for Java versions 7 (and below).

@eigenhombre reported this issue [earlier today](https://github.com/mcohen01/amazonica/pull/324#issuecomment-313513684) and since I'd like him (and his team) to be happy I've tried to turn this around quickly.

I haven't tested this on a box with Java 7 just yet but I should be able to just as soon as I get all of this VM business setup. 😼

### CODE REVIEW

**Comment by noprompt** [comment]
Oh yeah, it turned out reflection didn't work out the way I thought it might and I had to go with my back up plan of doing some lightweight `java.version` analysis. My original line of thought was that I could do some feature detection by trying to look for the `getParameters` method on `java.lang.reflect.Method`.

```clj
(some
 (fn [method]
   (= (.getName method)
      "getParameters"))
 (.getDeclaredMethods java.lang.reflect.Method))
```

This ended up returning `nil` and my hopes of determining the parameter name function in this manner were relegated to the pyre of my optimism. :skull_and_crossbones:

File: src/amazonica/core.clj [hunk 1]
```diff
@@ -921,41 +921,65 @@
...
   [method]
-  (let [parameters (. method getParameters)
-        names (map parameter-clojure-name parameters)
+  (let [names (parameter-names method)
         ;; This will help determine when parameter names should be
         ;; suffixed with an index i.e. `parameter-1`. Suffixing is
         ;; necessary when parameter names are synthesized from their
         ;; type names and the likelihood duplicates is high.
         name-frequency (frequencies names)]
-    (loop [names (map parameter-clojure-name parameters)
```
Comment (noprompt) [inline src/amazonica/core.clj]: Thank goodness for laziness! We could've been doing twice the amount of work here! 😸

**Comment by noprompt** [comment]
Sorry for the awkward branch name. I messed it up!

**Comment by eigenhombre** [comment]
Thanks for the quick work @noprompt !

**Comment by noprompt** [comment]
I was able to verify this works with at least Java 6.

```
user=> (System/getProperty "java.version")
"1.6.0_41"
user=> (require 'amazonica.core)
nil
user=> (require 'amazonica.aws.s3)
nil
user=> (:arglists (meta #'amazonica.aws.s3/get-object))
([get-object-request] [get-object-request file] [string-1 string-2])
```

@eigenhombre You bet!

**Comment by mcohen01** [comment]
sorry @eigenhombre, been on vacation. `0.3.107` is on clojars

**Comment by eigenhombre** [comment]
This one looks good for far -- thanks @mcohen01 @noprompt !

**Comment by noprompt** [comment]
@eigenhombre Awesome! 👍

**Comment by noprompt** [comment]
@mcohen01 No worries! Thanks!

**Pull request merged**